algorithms/linfa-logistic/src/lib.rs

```diff
@@ -273,46 +273,54 @@ where
     T: AsSingleTargets<Elem = C>,
     C: Ord + Clone,
 {
-    let y_single_target = y.as_single_targets();
-    let mut classes: Vec<&C> = vec![];
-    let mut target_vec = vec![];
-    let mut use_negative_label: bool = true;
-    for item in y_single_target {
-        if let Some(last_item) = classes.last() {
-            if *last_item != item {
-                use_negative_label = !use_negative_label;
+    let y = y.as_single_targets();
+
+    // find binary classes of our target dataset
+    let classes = y.iter().fold(Some((None, None)), |map, elm| match map {
+        Some((Some((n, val)), x)) if n == elm => Some((Some((n, val + 1)), x)),
+        Some((x, Some((n, val)))) if n == elm => Some((x, Some((n, val + 1)))),
+        Some((Some(x), None)) => Some((Some(x), Some((elm, 1)))),
+        Some((None, None)) => Some((Some((elm, 1)), None)),
```

Comment:
I think we should make this 2-tuple into a simple struct with named fields, so that this code is more understandable.